### PR TITLE
New/privacy check score

### PR DESCRIPTION
**Pull request by jreluiz**
PR regarding privacy rules

### CODE REVIEW

File: development/src/main/java/eubr/atmosphere/tma/planning/AdaptationManager.java [hunk 1]
```diff
@@ -6,21 +6,117 @@
...
+import eubr.atmosphere.tma.utils.PrivacyScore;
 
 public class AdaptationManager {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(AdaptationManager.class);
+    
     private static PlanManager planManager = new PlanManager();
+    private static ConfigRulesManager configRulesManager = new ConfigRulesManager();
 
+    public static void increasePRIVAASAnonymization(PrivacyScore privacyScore) {
```
Comment (joseadp) [inline development/src/main/java/eubr/atmosphere/tma/planning/AdaptationManager.java]: Não entendo o porquê disso tudo aqui. Não deveria estar no arquivo do Drools?

File: development/src/main/java/eubr/atmosphere/tma/planning/database/PlanManager.java [hunk 2]
```diff
@@ -1,26 +1,57 @@
...
 import eubr.atmosphere.tma.data.ConfigurationData;
 import eubr.atmosphere.tma.data.Plan;
+import eubr.atmosphere.tma.data.PlanStatus;
 import eubr.atmosphere.tma.utils.DatabaseManager;
 
 public class PlanManager {
     
     private static final Logger LOGGER = LoggerFactory.getLogger(PlanManager.class);
 
+    public Plan searchPlan(Integer planId) {
```
Comment (joseadp) [inline development/src/main/java/eubr/atmosphere/tma/planning/database/PlanManager.java]: Para quê este método é preciso?

File: development/src/main/java/eubr/atmosphere/tma/planning/AdaptationManager.java (hunk 1 above)
Comment (jreluiz) [reply development/src/main/java/eubr/atmosphere/tma/planning/AdaptationManager.java]: Esse método é chamado do arquivo do drools. Se não estiver de acordo com as definições de vocês fiquem à vontade para alterar.

File: development/src/main/java/eubr/atmosphere/tma/planning/database/PlanManager.java (hunk 2 above)
Comment (jreluiz) [reply development/src/main/java/eubr/atmosphere/tma/planning/database/PlanManager.java]: Esse método seria usado para buscar o plano executado no ciclo anterior do MAPE (uma ideia que tivemos de compor o plano a medida que o ciclo é executado),  porém acho que não vai funcionar assim, então pode remover.

@joseadp Acho mais fácil vc usar o arquivo de regras e adequar o código de acordo com o que vcs fizeram (composição do plano, etc), senão vamos gastar muito tempo tentando sincronizar isso.

File: development/src/main/java/eubr/atmosphere/tma/planning/AdaptationManager.java (hunk 1 above)
Comment (nmsa) [reply development/src/main/java/eubr/atmosphere/tma/planning/AdaptationManager.java]: > "Se não estiver de acordo com as definições de vocês fiquem à vontade para alterar."

Bom, a gente pode alterar, mas não é esse o espírito da review.
A ideia é dar feedback aos developers e ajudar a manter a uniformidade no projecto. 
Não é muito positivo cada um fazer como quer e depois o resto que se preocupe em alterar para uniformizar.

/cc @brenofranca83 @joseadp @mpvieira

**Comment by brenofranca83** [comment]
I believe now it is ok, could you please check @joseadp ?

**Comment by joseadp** [comment]
@brenofranca83 I believe it is ok.
My only question is regarding the Kafka queue. Aren't you using it to Analyze send a message when a new score is calculated?
(cc @nmsa )

**Comment by brenofranca83** [comment]
> @brenofranca83 I believe it is ok.
> My only question is regarding the Kafka queue. Aren't you using it to Analyze send a message when a new score is calculated?
> (cc @nmsa )

@jreluiz Are we?

**Comment by jreluiz** [comment]
> > @brenofranca83 I believe it is ok.
> > My only question is regarding the Kafka queue. Aren't you using it to Analyze send a message when a new score is calculated?
> > (cc @nmsa )
> 
> @jreluiz Are we?

Yes, we are using the kafka queue "topic-planning"to send a Analyze message to the Planning.

**Comment by joseadp** [comment]
> Yes, we are using the kafka queue "topic-planning"to send a Analyze message to the Planning.

@jreluiz , I have just noticed that you are using kafka queue. I'm in doubt about we have two KieSession objects.
The idea is to have only one, and apply all the rules (either of performance or privacy). If the object is in the KieSession is of a different type, that rule will not match the "when" condition.
Couldn't we simplify that and have only one KieSession? Additionally, we have one of them commented, which is not a good practice.
(cc @brenofranca83 @nmsa )

**Comment by jreluiz** [comment]
> > Yes, we are using the kafka queue "topic-planning"to send a Analyze message to the Planning.
> 
> @jreluiz , I have just noticed that you are using kafka queue. I'm in doubt about we have two KieSession objects.
> The idea is to have only one, and apply all the rules (either of performance or privacy). If the object is in the KieSession is of a different type, that rule will not match the "when" condition.
> Couldn't we simplify that and have only one KieSession? Additionally, we have one of them commented, which is not a good practice.
> (cc @brenofranca83 @nmsa )

@joseadp I tried to leave only one KieSession object, but the objects with data coming from Analyze are different for privacy and performance. Also, for privacy, we need to fetch from the base the threshold value that will be used in the drools rule. Regarding the commented code, thanks for warning: as was testing only privacy I left the performance comment, I will remove the comment. Feel free to try leaving only one KieSession object, I couldn't.. (cc @brenofranca83 @nmsa)

**Comment by joseadp** [comment]
> @joseadp I tried to leave only one KieSession object, but the objects with data coming from Analyze are different for privacy and performance. Also, for privacy, we need to fetch from the base the threshold value that will be used in the drools rule. Regarding the commented code, thanks for warning: as was testing only privacy I left the performance comment, I will remove the comment. Feel free to try leaving only one KieSession object, I couldn't.. (cc @brenofranca83 @nmsa)

@jreluiz , it is not an issue to have different types of objects in the same session. If you check any of the Drools files (for instance https://github.com/eubr-atmosphere/tma-framework-p/blob/e222b16e1d946a00fe536bd154b9aa8bb3fab042/development/src/main/resources/eubr/atmosphere/tma/planning/privacy/PrivacyRules.drl), you will see that the `when` condition checks if the object is of a specific type (in this example `PrivacyScore`). If the object is of a different type, the rule will not match. As all of the rules are applied, for the rules that have a different `when` condition, the other object may have an adaptation dispatched by that rule.
Hence, it is not necessary to have different KieSession objects.

**Comment by jreluiz** [comment]
> > @joseadp I tried to leave only one KieSession object, but the objects with data coming from Analyze are different for privacy and performance. Also, for privacy, we need to fetch from the base the threshold value that will be used in the drools rule. Regarding the commented code, thanks for warning: as was testing only privacy I left the performance comment, I will remove the comment. Feel free to try leaving only one KieSession object, I couldn't.. (cc @brenofranca83 @nmsa)
> 
> @jreluiz , it is not an issue to have different types of objects in the same session. If you check any of the Drools files (for instance https://github.com/eubr-atmosphere/tma-framework-p/blob/e222b16e1d946a00fe536bd154b9aa8bb3fab042/development/src/main/resources/eubr/atmosphere/tma/planning/privacy/PrivacyRules.drl), you will see that the `when` condition checks if the object is of a specific type (in this example `PrivacyScore`). If the object is of a different type, the rule will not match. As all of the rules are applied, for the rules that have a different `when` condition, the other object may have an adaptation dispatched by that rule.
> Hence, it is not necessary to have different KieSession objects.

@joseadp, I think I understand. I made some modifications according to what you requested, but when testing the planning I got the following error:

`[INFO ] 01:09:40.604 eubr.atmosphere.tma.planning.Main.runConsumer() - ConsumerRecords: 1
[INFO ] 01:09:40.645 eubr.atmosphere.tma.planning.Main.validateValue() - ConsumerRecord(topic = topic-planning, partition = 0, offset = 7, CreateTime = 1568768977109, serialized key size = 8, serialized value size = 89, headers = RecordHeaders(headers = [], isReadOnly = false), key = 1568768976863, value = {"configurationProfileId":1,"attributeId":1,"score":0.47708418060831753,"threshold":0.05})
[INFO ] 01:09:40.645 eubr.atmosphere.tma.planning.Main.validateValue() - Score: 0.47708418060831753 / Offset: 7
Exception in thread "main" java.lang.RuntimeException: Error evaluating constraint 'privacyScore.score < privacyScore.getThreshold()' in [Rule "Privacy Score Validation - Increase PRIVAAS Anonymization" in Privacy.drl]
	at org.drools.core.rule.constraint.MvelConstraint.evaluate(MvelConstraint.java:274)
	at org.drools.core.rule.constraint.MvelConstraint.isAllowed(MvelConstraint.java:222)
	at org.drools.core.reteoo.AlphaNode.assertObject(AlphaNode.java:139)
	at org.drools.core.reteoo.CompositeObjectSinkAdapter.doPropagateAssertObject(CompositeObjectSinkAdapter.java:498)
	at org.drools.core.reteoo.CompositeObjectSinkAdapter.propagateAssertObject(CompositeObjectSinkAdapter.java:396)
	at org.drools.core.reteoo.ObjectTypeNode.propagateAssert(ObjectTypeNode.java:323)
	at org.drools.core.phreak.PropagationEntry$Insert.propagate(PropagationEntry.java:161)
	at org.drools.core.phreak.PropagationEntry$Insert.execute(PropagationEntry.java:166)
	at org.drools.core.phreak.SynchronizedPropagationList.flush(SynchronizedPropagationList.java:96)
	at org.drools.core.phreak.SynchronizedPropagationList.flush(SynchronizedPropagationList.java:91)
	at org.drools.core.common.DefaultAgenda.fireLoop(DefaultAgenda.java:1051)
	at org.drools.core.common.DefaultAgenda.internalFireAllRules(DefaultAgenda.java:1014)
	at org.drools.core.common.DefaultAgenda.fireAllRules(DefaultAgenda.java:1006)
	at org.drools.core.impl.StatefulKnowledgeSessionImpl.internalFireAllRules(StatefulKnowledgeSessionImpl.java:1321)
	at org.drools.core.impl.StatefulKnowledgeSessionImpl.fireAllRules(StatefulKnowledgeSessionImpl.java:1312)
	at org.drools.core.impl.StatefulKnowledgeSessionImpl.fireAllRules(StatefulKnowledgeSessionImpl.java:1296)
	at eubr.atmosphere.tma.planning.Main.runConsumer(Main.java:72)
	at eubr.atmosphere.tma.planning.Main.main(Main.java:39)
Caused by: [Error: null pointer: privacyScore.score]
[Near : {... privacyScore.score < privacySc ....}]`

Apparently the TrustworthinessScore object (privacyScore) is null. I couldn't understand because the object exists, is not null and is being passed to the rule.
Any suggestion?

**Comment by joseadp** [comment]
> > > @joseadp I tried to leave only one KieSession object, but the objects with data coming from Analyze are different for privacy and performance. Also, for privacy, we need to fetch from the base the threshold value that will be used in the drools rule. Regarding the commented code, thanks for warning: as was testing only privacy I left the performance comment, I will remove the comment. Feel free to try leaving only one KieSession object, I couldn't.. (cc @brenofranca83 @nmsa)
> > 
> > 
> > @jreluiz , it is not an issue to have different types of objects in the same session. If you check any of the Drools files (for instance https://github.com/eubr-atmosphere/tma-framework-p/blob/e222b16e1d946a00fe536bd154b9aa8bb3fab042/development/src/main/resources/eubr/atmosphere/tma/planning/privacy/PrivacyRules.drl), you will see that the `when` condition checks if the object is of a specific type (in this example `PrivacyScore`). If the object is of a different type, the rule will not match. As all of the rules are applied, for the rules that have a different `when` condition, the other object may have an adaptation dispatched by that rule.
> > Hence, it is not necessary to have different KieSession objects.
> 
> @joseadp, I think I understand. I made some modifications according to what you requested, but when testing the planning I got the following error:
> 
> `[INFO ] 01:09:40.604 eubr.atmosphere.tma.planning.Main.runConsumer() - ConsumerRecords: 1 [INFO ] 01:09:40.645 eubr.atmosphere.tma.planning.Main.validateValue() - ConsumerRecord(topic = topic-planning, partition = 0, offset = 7, CreateTime = 1568768977109, serialized key size = 8, serialized value size = 89, headers = RecordHeaders(headers = [], isReadOnly = false), key = 1568768976863, value = {"configurationProfileId":1,"attributeId":1,"score":0.47708418060831753,"threshold":0.05}) [INFO ] 01:09:40.645 eubr.atmosphere.tma.planning.Main.validateValue() - Score: 0.47708418060831753 / Offset: 7 Exception in thread "main" java.lang.RuntimeException: Error evaluating constraint 'privacyScore.score < privacyScore.getThreshold()' in [Rule "Privacy Score Validation - Increase PRIVAAS Anonymization" in Privacy.drl] at org.drools.core.rule.constraint.MvelConstraint.evaluate(MvelConstraint.java:274) at org.drools.core.rule.constraint.MvelConstraint.isAllowed(MvelConstraint.java:222) at org.drools.core.reteoo.AlphaNode.assertObject(AlphaNode.java:139) at org.drools.core.reteoo.CompositeObjectSinkAdapter.doPropagateAssertObject(CompositeObjectSinkAdapter.java:498) at org.drools.core.reteoo.CompositeObjectSinkAdapter.propagateAssertObject(CompositeObjectSinkAdapter.java:396) at org.drools.core.reteoo.ObjectTypeNode.propagateAssert(ObjectTypeNode.java:323) at org.drools.core.phreak.PropagationEntry$Insert.propagate(PropagationEntry.java:161) at org.drools.core.phreak.PropagationEntry$Insert.execute(PropagationEntry.java:166) at org.drools.core.phreak.SynchronizedPropagationList.flush(SynchronizedPropagationList.java:96) at org.drools.core.phreak.SynchronizedPropagationList.flush(SynchronizedPropagationList.java:91) at org.drools.core.common.DefaultAgenda.fireLoop(DefaultAgenda.java:1051) at org.drools.core.common.DefaultAgenda.internalFireAllRules(DefaultAgenda.java:1014) at org.drools.core.common.DefaultAgenda.fireAllRules(DefaultAgenda.java:1006) at org.drools.core.impl.StatefulKnowledgeSessionImpl.internalFireAllRules(StatefulKnowledgeSessionImpl.java:1321) at org.drools.core.impl.StatefulKnowledgeSessionImpl.fireAllRules(StatefulKnowledgeSessionImpl.java:1312) at org.drools.core.impl.StatefulKnowledgeSessionImpl.fireAllRules(StatefulKnowledgeSessionImpl.java:1296) at eubr.atmosphere.tma.planning.Main.runConsumer(Main.java:72) at eubr.atmosphere.tma.planning.Main.main(Main.java:39) Caused by: [Error: null pointer: privacyScore.score] [Near : {... privacyScore.score < privacySc ....}]`
> 
> Apparently the TrustworthinessScore object (privacyScore) is null. I couldn't understand because the object exists, is not null and is being passed to the rule.
> Any suggestion?

@jreluiz , thank you for addressing this item!
The issue that you are facing is related to the way that you are accessing the object.
Different from regular Java, when you are using Drools, you need to access the attributes instead of the class methods. See this example: https://github.com/eubr-atmosphere/tma-framework-p/blob/master/development/src/main/resources/eubr/atmosphere/tma/planning/WSVDDemoPython.drl
Hence, you need to change `.getThreshold()` to `threshold` in the drools file (https://github.com/eubr-atmosphere/tma-framework-p/pull/9/files#diff-2ce7c4b1c2e62b9fa96b5ce6b8e7bc98).
Let me know if it works.

**Comment by jreluiz** [comment]
> > > > @joseadp I tried to leave only one KieSession object, but the objects with data coming from Analyze are different for privacy and performance. Also, for privacy, we need to fetch from the base the threshold value that will be used in the drools rule. Regarding the commented code, thanks for warning: as was testing only privacy I left the performance comment, I will remove the comment. Feel free to try leaving only one KieSession object, I couldn't.. (cc @brenofranca83 @nmsa)
> > > 
> > > 
> > > @jreluiz , it is not an issue to have different types of objects in the same session. If you check any of the Drools files (for instance https://github.com/eubr-atmosphere/tma-framework-p/blob/e222b16e1d946a00fe536bd154b9aa8bb3fab042/development/src/main/resources/eubr/atmosphere/tma/planning/privacy/PrivacyRules.drl), you will see that the `when` condition checks if the object is of a specific type (in this example `PrivacyScore`). If the object is of a different type, the rule will not match. As all of the rules are applied, for the rules that have a different `when` condition, the other object may have an adaptation dispatched by that rule.
> > > Hence, it is not necessary to have different KieSession objects.
> > 
> > 
> > @joseadp, I think I understand. I made some modifications according to what you requested, but when testing the planning I got the following error:
> > `[INFO ] 01:09:40.604 eubr.atmosphere.tma.planning.Main.runConsumer() - ConsumerRecords: 1 [INFO ] 01:09:40.645 eubr.atmosphere.tma.planning.Main.validateValue() - ConsumerRecord(topic = topic-planning, partition = 0, offset = 7, CreateTime = 1568768977109, serialized key size = 8, serialized value size = 89, headers = RecordHeaders(headers = [], isReadOnly = false), key = 1568768976863, value = {"configurationProfileId":1,"attributeId":1,"score":0.47708418060831753,"threshold":0.05}) [INFO ] 01:09:40.645 eubr.atmosphere.tma.planning.Main.validateValue() - Score: 0.47708418060831753 / Offset: 7 Exception in thread "main" java.lang.RuntimeException: Error evaluating constraint 'privacyScore.score < privacyScore.getThreshold()' in [Rule "Privacy Score Validation - Increase PRIVAAS Anonymization" in Privacy.drl] at org.drools.core.rule.constraint.MvelConstraint.evaluate(MvelConstraint.java:274) at org.drools.core.rule.constraint.MvelConstraint.isAllowed(MvelConstraint.java:222) at org.drools.core.reteoo.AlphaNode.assertObject(AlphaNode.java:139) at org.drools.core.reteoo.CompositeObjectSinkAdapter.doPropagateAssertObject(CompositeObjectSinkAdapter.java:498) at org.drools.core.reteoo.CompositeObjectSinkAdapter.propagateAssertObject(CompositeObjectSinkAdapter.java:396) at org.drools.core.reteoo.ObjectTypeNode.propagateAssert(ObjectTypeNode.java:323) at org.drools.core.phreak.PropagationEntry$Insert.propagate(PropagationEntry.java:161) at org.drools.core.phreak.PropagationEntry$Insert.execute(PropagationEntry.java:166) at org.drools.core.phreak.SynchronizedPropagationList.flush(SynchronizedPropagationList.java:96) at org.drools.core.phreak.SynchronizedPropagationList.flush(SynchronizedPropagationList.java:91) at org.drools.core.common.DefaultAgenda.fireLoop(DefaultAgenda.java:1051) at org.drools.core.common.DefaultAgenda.internalFireAllRules(DefaultAgenda.java:1014) at org.drools.core.common.DefaultAgenda.fireAllRules(DefaultAgenda.java:1006) at org.drools.core.impl.StatefulKnowledgeSessionImpl.internalFireAllRules(StatefulKnowledgeSessionImpl.java:1321) at org.drools.core.impl.StatefulKnowledgeSessionImpl.fireAllRules(StatefulKnowledgeSessionImpl.java:1312) at org.drools.core.impl.StatefulKnowledgeSessionImpl.fireAllRules(StatefulKnowledgeSessionImpl.java:1296) at eubr.atmosphere.tma.planning.Main.runConsumer(Main.java:72) at eubr.atmosphere.tma.planning.Main.main(Main.java:39) Caused by: [Error: null pointer: privacyScore.score] [Near : {... privacyScore.score < privacySc ....}]`
> > Apparently the TrustworthinessScore object (privacyScore) is null. I couldn't understand because the object exists, is not null and is being passed to the rule.
> > Any suggestion?
> 
> @jreluiz , thank you for addressing this item!
> The issue that you are facing is related to the way that you are accessing the object.
> Different from regular Java, when you are using Drools, you need to access the attributes instead of the class methods. See this example: https://github.com/eubr-atmosphere/tma-framework-p/blob/master/development/src/main/resources/eubr/atmosphere/tma/planning/WSVDDemoPython.drl
> Hence, you need to change `.getThreshold()` to `threshold` in the drools file (https://github.com/eubr-atmosphere/tma-framework-p/pull/9/files#diff-2ce7c4b1c2e62b9fa96b5ce6b8e7bc98).
> Let me know if it works.

@joseadp, thank you for the tips.
I did what you suggested but the same error still occurs, apparently a NullPointer on the line: `privacyScore.score < privacyScore.threshold`

.drl file with updated rules: [https://github.com/eubr-atmosphere/tma-framework-p/blob/new/privacy_check_score/development/src/main/resources/eubr/atmosphere/tma/planning/Privacy.drl](url)

Any suggestion?

**Comment by joseadp** [comment]
> @joseadp, thank you for the tips.
> I did what you suggested but the same error still occurs, a NullPointer on the line:
> $score: TrustworthinessScore ( **privacyScore**.score < privacyScore.threshold )
> -> Apparently the NullPointer occurs in the privacyScore attribute.
> 
> .drl file with updated rules: [https://github.com/eubr-atmosphere/tma-framework-p/blob/new/privacy_check_score/development/src/main/resources/eubr/atmosphere/tma/planning/Privacy.drl](url)
> 
> Any suggestion?

@jreluiz , my suggestion is to debug the rules.
As the message says that the privacyScore is null, you should try to validate if it is null or not.
Try removing the items to check that. For instance, removes the condition `privacyScore.score < privacyScore.threshold`, and try to check if the `then` block is executed. In the `then` block, you should try to print to the output the TrustworthinessScore object, and its attributes, in order to detect where is the null object that is being reported.

**Comment by jreluiz** [comment]
> > @joseadp, thank you for the tips.
> > I did what you suggested but the same error still occurs, a NullPointer on the line:
> > $score: TrustworthinessScore ( **privacyScore**.score < privacyScore.threshold )
> > -> Apparently the NullPointer occurs in the privacyScore attribute.
> > .drl file with updated rules: [https://github.com/eubr-atmosphere/tma-framework-p/blob/new/privacy_check_score/development/src/main/resources/eubr/atmosphere/tma/planning/Privacy.drl](url)
> > Any suggestion?
> 
> @jreluiz , my suggestion is to debug the rules.
> As the message says that the privacyScore is null, you should try to validate if it is null or not.
> Try removing the items to check that. For instance, removes the condition `privacyScore.score < privacyScore.threshold`, and try to check if the `then` block is executed. In the `then` block, you should try to print to the output the TrustworthinessScore object, and its attributes, in order to detect where is the null object that is being reported.

@joseadp, I made some modifications according to your suggestions and needed to change the TrustworthinessScore class by adding some attributes. I did tests and it worked fine, please validate.

**Comment by joseadp** [comment]
> @joseadp, I made some modifications according to your suggestions and needed to change the TrustworthinessScore class by adding some attributes. I did tests and it worked fine, please validate.

I didn't get why you removed the privacy score from the TrustworthinessScore. Both the score and threshold (as it is developed) reflect to Trustworthiness. From my understanding, your rules are for Privacy.
Additionally, the TrustworthinessScore class does not have some of the attributes that your are accessing (I am checking this PR: https://github.com/eubr-atmosphere/tma-framework/pull/16/files - if you have other updates, they are not committed). I am not sure how this is working.

**Comment by jreluiz** [comment]
> > @joseadp, I made some modifications according to your suggestions and needed to change the TrustworthinessScore class by adding some attributes. I did tests and it worked fine, please validate.
> 
> I didn't get why you removed the privacy score from the TrustworthinessScore. Both the score and threshold (as it is developed) reflect to Trustworthiness. From my understanding, your rules are for Privacy.
> Additionally, the TrustworthinessScore class does not have some of the attributes that your are accessing (I am checking this PR: https://github.com/eubr-atmosphere/tma-framework/pull/16/files - if you have other updates, they are not committed). I am not sure how this is working.

@joseadp , sorry, I really had not committed changes to the tma-utils project. I just committed..
In my tests the line `TrustworthinessScore score = new Gson (). FromJson (stringJsonScore, TrustworthinessScore.class);` does not populate the PrivacyScore object that is within the TrustworthinessScore class, so the only way I found it was to move the PrivacyScore attributes to TrustworthinessScore. 

What do you think? Thanks for the help.

**Comment by joseadp** [comment]
> @joseadp , sorry, I really had not committed changes to the tma-utils project. I just committed..
> In my tests the line `TrustworthinessScore score = new Gson (). FromJson (stringJsonScore, TrustworthinessScore.class);` does not populate the PrivacyScore object that is within the TrustworthinessScore class, so the only way I found it was to move the PrivacyScore attributes to TrustworthinessScore.
> 
> What do you think? Thanks for the help.

@jreluiz , like I mentioned in my previous message, I believe this does not make sense. We can have score on the trustworthiness level too, so it gets mixed with the other score and threshold.
Can you point where is this line from your tests? I couldn't find in the PRs.

**Comment by jreluiz** [comment]
> TrustworthinessScore score = new Gson (). FromJson (stringJsonScore, TrustworthinessScore.class);

@joseadp, this line is in the Main.java class [https://github.com/eubr-atmosphere/tma-framework-p/blob/master/development/src/main/java/eubr/atmosphere/tma/planning/Main.java](url) method validateValue:

`private static void validateValue(ConsumerRecord<Long, String> record, KieSession ksession) {
        String stringJsonScore = record.value();
        TrustworthinessScore score = new Gson().fromJson(stringJsonScore, TrustworthinessScore.class);
        LOGGER.info(record.toString());
        LOGGER.info("Score: {} / Offset: {}", score.getScore(), record.offset());
        factHandleList.add(ksession.insert(score));
    }`
As commented, this method does not populate the declared PrivacyScore object within TrustworthinessScore.

I saw that you declared the podCount attribute in the TrustworthinessScore class, I think for the same reason:

`private Integer podCount;`

If I did something like:
`PrivacyScore score = new Gson().fromJson(stringJsonScore, Privacy.class);`
I think it would work, but it's not generic, right? And that's not the intention..

**Comment by joseadp** [comment]
> > TrustworthinessScore score = new Gson (). FromJson (stringJsonScore, TrustworthinessScore.class);
> 
> @joseadp, this line is in the Main.java class [https://github.com/eubr-atmosphere/tma-framework-p/blob/master/development/src/main/java/eubr/atmosphere/tma/planning/Main.java](url) method validateValue:
> 
> `private static void validateValue(ConsumerRecord<Long, String> record, KieSession ksession) { String stringJsonScore = record.value(); TrustworthinessScore score = new Gson().fromJson(stringJsonScore, TrustworthinessScore.class); LOGGER.info(record.toString()); LOGGER.info("Score: {} / Offset: {}", score.getScore(), record.offset()); factHandleList.add(ksession.insert(score)); }`
> As commented, this method does not populate the declared PrivacyScore object within TrustworthinessScore.
> 
> I saw that you declared the podCount attribute in the TrustworthinessScore class, I think for the same reason:
> 
> `private Integer podCount;`
> 
> If I did something like:
> `PrivacyScore score = new Gson().fromJson(stringJsonScore, Privacy.class);`
> I think it would work, but it's not generic, right? And that's not the intention..

@jreluiz , the reason is different. The initial TrustworthinessScore was more related to `ResourceConsumptionPerPod`. It also needs some refactoring.
Nevertheless, the problem is different.

I am assuming you are using the same way to parse the TrustworthinessScore object (line 86 from this file: https://github.com/eubr-atmosphere/tma-framework-p/blob/master/development/src/main/java/eubr/atmosphere/tma/planning/Main.java).
This should load all the attributes (including objects) from the TrustworthinessScore object. If this is not happening, this is not something on Drools side. It should be something else.
Have you tried to print this object on the Main class to check if the object is null? Maybe you are using an outdated version of tma-utils. Can you double-check it?

**Comment by jreluiz** [comment]
> > > TrustworthinessScore score = new Gson (). FromJson (stringJsonScore, TrustworthinessScore.class);
> > 
> > 
> > @joseadp, this line is in the Main.java class [https://github.com/eubr-atmosphere/tma-framework-p/blob/master/development/src/main/java/eubr/atmosphere/tma/planning/Main.java](url) method validateValue:
> > `private static void validateValue(ConsumerRecord<Long, String> record, KieSession ksession) { String stringJsonScore = record.value(); TrustworthinessScore score = new Gson().fromJson(stringJsonScore, TrustworthinessScore.class); LOGGER.info(record.toString()); LOGGER.info("Score: {} / Offset: {}", score.getScore(), record.offset()); factHandleList.add(ksession.insert(score)); }`
> > As commented, this method does not populate the declared PrivacyScore object within TrustworthinessScore.
> > I saw that you declared the podCount attribute in the TrustworthinessScore class, I think for the same reason:
> > `private Integer podCount;`
> > If I did something like:
> > `PrivacyScore score = new Gson().fromJson(stringJsonScore, Privacy.class);`
> > I think it would work, but it's not generic, right? And that's not the intention..
> 
> @jreluiz , the reason is different. The initial TrustworthinessScore was more related to `ResourceConsumptionPerPod`. It also needs some refactoring.
> Nevertheless, the problem is different.
> 
> I am assuming you are using the same way to parse the TrustworthinessScore object (line 86 from this file: https://github.com/eubr-atmosphere/tma-framework-p/blob/master/development/src/main/java/eubr/atmosphere/tma/planning/Main.java).
> This should load all the attributes (including objects) from the TrustworthinessScore object. If this is not happening, this is not something on Drools side. It should be something else.
> Have you tried to print this object on the Main class to check if the object is null? Maybe you are using an outdated version of tma-utils. Can you double-check it?

PrivacyScore object is null: 
```
[INFO ] 23:07:53.820 eubr.atmosphere.tma.planning.Main.validateValue() - privacyScore: null
[INFO ] 23:07:53.820 eubr.atmosphere.tma.planning.Main.validateValue() - ConfigurationProfileID: null
[INFO ] 23:07:53.821 eubr.atmosphere.tma.planning.Main.validateValue() - AttributeID: null
[INFO ] 23:07:53.821 eubr.atmosphere.tma.planning.Main.validateValue() - Threshold: null
```
The json format being sent by Analyzer is correct:
`value = {"configurationProfileId":1,"attributeId":1,"score":0.47708418060831753,"threshold":0.05}`

Are you sure that ResourceConsumptionScore and PerformanceScore objects are being populated using:
`TrustworthinessScore score = new Gson().fromJson(stringJsonScore, TrustworthinessScore.class);`

**Comment by joseadp** [comment]
> Are you sure that ResourceConsumptionScore and PerformanceScore objects are being populated using:
> `TrustworthinessScore score = new Gson().fromJson(stringJsonScore, TrustworthinessScore.class);`

@jreluiz yes.
If the object that analyze is sending is not correct. From the JSON that you sent, not all the information is available.
From our deprecated version of analyze (https://github.com/joseadp/tma-analyze), we convert the TrustworthinessScore object using the following command:
`JsonElement jsonElement = new Gson().toJsonTree(score);`
Are you using the same one? From the version that we have on master branch from Analyze (https://github.com/eubr-atmosphere/tma-framework-a/blob/master/analyzer/src/main/java/eubrazil/atmosphere/kafka/KafkaManager.java), a different object is being used to be added on Kafka topic.

Please, check it. There are too many branches on tma-framework-a repository, and it is hard to review them all.

**Comment by jreluiz** [comment]
> > Are you sure that ResourceConsumptionScore and PerformanceScore objects are being populated using:
> > `TrustworthinessScore score = new Gson().fromJson(stringJsonScore, TrustworthinessScore.class);`
> 
> @jreluiz yes.
> If the object that analyze is sending is not correct. From the JSON that you sent, not all the information is available.
> From our deprecated version of analyze (https://github.com/joseadp/tma-analyze), we convert the TrustworthinessScore object using the following command:
> `JsonElement jsonElement = new Gson().toJsonTree(score);`
> Are you using the same one? From the version that we have on master branch from Analyze (https://github.com/eubr-atmosphere/tma-framework-a/blob/master/analyzer/src/main/java/eubrazil/atmosphere/kafka/KafkaManager.java), a different object is being used to be added on Kafka topic.
> 
> Please, check it. There are too many branches on tma-framework-a repository, and it is hard to review them all.

@joseadp, That's right. We were sending the PrivacyScore object in Analyze, I changed it to TrustworthinessScore and everything worked perfectly. 
The Analyze version we are using is feat_analyzer_integration, but we will probably have another version after knowledge unification. 

Thanks a lot for the help. Any questions or problems, I am available.

**Comment by joseadp** [comment]
> The Analyze version we are using is feat_analyzer_integration, but we will probably have another version after knowledge unification.

@jreluiz Great! If there is any other branch in that repository to be deleted (the ones that were already merged), please delete them. It will help us with that.
For this PR, I will approve it, and then we can merge it.

@nmsa @brenofranca83 , do you agree with that?

**Pull request merged**